### ISSUE TITLE
Reduce VPN Pod Privileges

### ISSUE THREAD

**Issue by marwinski** [gardener/vpn#41]
The VPN pods (both seed and shoot) run with ```privileged: true``` and the ```NET_ADMIN``` capability. We should try to reduce that to the minimum required privileges; possibly by using init containers.

This said, this might prove hard to do for the VPN pods as openvpn needs to create the tun interfaces but maybe we could get rid of the ```privileged``` setting.

**Comment by vlerenc** [comment]
@zanetworker @DockToFuture @mvladev Please forgive brevity while grooming the backlog. Issue still relevant in the context of the new VPN solution or would you like to close it (it is rotten)?

**Comment by DockToFuture** [comment]
This will be obsolete with the new "konnectivity" setup (https://github.com/gardener/gardener/pull/2251) for cluster >= 1.18, shall we leave it open, because its still relevant for cluster with lower k8s versions? @mvladev @zanetworker ?

**Comment by zanetworker** [comment]
I don't think we can get rid of the `priviliged` option since this component creates network interfaces on the node.

**Comment by vlerenc** [comment]
...then we could close the ticket @zanetworker. But please also see the proposal by @marwinski : init containers. Then again, init containers have some drawbacks (not being rerun), but maybe OK in this case. If in the end, moving it to init containers has no value (but I would assume it would improve security posture, although I can't assess by how much/whether it's worth it) or getting rid of the permissions isn't possible, then we can close this one.

**Comment by ThormaehlenFred** [comment]
Hi @vlerenc , this issue should be processed in Q3 before the next pentest, which is planned for end of September, beginning of October. And it needs an assignee. How should we proceed?

**Comment by DockToFuture** [comment]
The privileged mode is required for setting the kernel parameter in https://github.com/gardener/vpn2/blob/79b0dd2812f62ede66245a39bc03f12893004db4/shoot-client/network-connection.sh#L25-L28. I'm asking myself however what was the reason that they were set to those specific values as they correspond to the default values from https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/net/tcp.h#n155. Were they somewhere overwritten in one of our supported os's? Otherwise we could remove this section at all.

**Comment by ThormaehlenFred** [comment]
FYI, I am reporting this issue for multiple months already in each RDM. A decision how to proceed would be very much appreciated.

**Comment by ialidzhikov** [comment]
@vlerenc can we move this issue to the https://github.com/gardener/vpn2 repo? I guess we don't want to do anything anymore for the old https://github.com/gardener/vpn and fix it only for https://github.com/gardener/vpn2, right?

**Comment by marwinski** [comment]
I have looked into this but got preempted by another issue. This said, the kernel parameters can be set in an init container but OpenVPN remains difficult. So far I have not found the required capabilities to run it non-privileged. Just CAP_NET_ADMIN and CAP_NET_RAW will not do the trick.

**Comment by vlerenc** [comment]
How about binary-search? I launch my ops-pod with the full list (never bothered to trim the list down):

```
        - SETPCAP
        - SYS_MODULE
        - SYS_RAWIO
        - SYS_PACCT
        - SYS_ADMIN
        - SYS_NICE
        - SYS_RESOURCE
        - SYS_TIME
        - SYS_TTY_CONFIG
        - MKNOD
        - AUDIT_WRITE
        - AUDIT_CONTROL
        - MAC_OVERRIDE
        - MAC_ADMIN
        - NET_ADMIN
        - SYSLOG
        - CHOWN
        - NET_RAW
        - DAC_OVERRIDE
        - FOWNER
        - DAC_READ_SEARCH
        - FSETID
        - KILL
        - SETGID
        - SETUID
        - LINUX_IMMUTABLE
        - NET_BIND_SERVICE
        - NET_BROADCAST
        - IPC_LOCK
        - IPC_OWNER
        - SYS_CHROOT
        - SYS_PTRACE
        - SYS_BOOT
        - LEASE
        - SETFCAP
        - WAKE_ALARM
        - BLOCK_SUSPEND
```

Anything there that rings a bell?

**Comment by marwinski** [comment]
I was debugging before I got preempted. I have straces on where it is failing but that particular place is not the root cause (tries to open a file that does not exist). It appears there is a failure earlier on that OpenVPN does not report...

**Comment by ialidzhikov** [comment]
For me in non-privileged mode the vpn-shoot container fails with:
```
2022-07-13T07:06:12.133178904Z stdout F 2022-07-13 07:06:12 DEPRECATED OPTION: --cipher set to 'AES-256-CBC' but missing in --data-ciphers (AES-256-GCM:AES-128-GCM). Future OpenVPN version will ignore --cipher for cipher negotiations. Add 'AES-256-CBC' to --data-ciphers or change --cipher 'AES-256-CBC' to --data-ciphers-fallback 'AES-256-CBC' to silence this warning.
2022-07-13T07:06:12.133364918Z stdout F 2022-07-13 07:06:12 WARNING: file '/srv/secrets/vpn-shoot/tls.key' is group or others accessible
2022-07-13T07:06:12.133406801Z stdout F 2022-07-13 07:06:12 OpenVPN 2.5.6 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Mar 24 2022
2022-07-13T07:06:12.13346157Z stdout F 2022-07-13 07:06:12 library versions: OpenSSL 1.1.1n  15 Mar 2022, LZO 2.10
2022-07-13T07:06:12.134889514Z stdout F 2022-07-13 07:06:12 TCP/UDP: Preserving recently used remote address: [AF_INET]108.128.8.82:8132
2022-07-13T07:06:12.134967723Z stdout F 2022-07-13 07:06:12 Attempting to establish TCP connection with [AF_INET]108.128.8.82:8132 [nonblock]
2022-07-13T07:06:12.136722821Z stdout F 2022-07-13 07:06:12 TCP connection established with [AF_INET]108.128.8.82:8132
2022-07-13T07:06:14.141174146Z stdout F 2022-07-13 07:06:14 TCP_CLIENT link local: (not bound)
2022-07-13T07:06:14.141232942Z stdout F 2022-07-13 07:06:14 TCP_CLIENT link remote: [AF_INET]108.128.8.82:8132
2022-07-13T07:06:14.148312942Z stdout F 2022-07-13 07:06:14 [vpn-seed-server] Peer Connection Initiated with [AF_INET]108.128.8.82:8132
2022-07-13T07:06:14.149777593Z stdout F 2022-07-13 07:06:14 ERROR: Cannot open TUN/TAP dev /dev/net/tun: No such file or directory (errno=2)
2022-07-13T07:06:14.149793133Z stdout F 2022-07-13 07:06:14 Exiting due to fatal error
```

It is not a capability issue. It fails the same way even if you add to the Pod all of the possible capabilities.
By default a container is not allowed to access any devices on the host, but a "privileged" container is given access to all devices on the host.

@marwinski I assume this is the same problem that you faced as well?

**Comment by marwinski** [comment]
precisely, but this is not the root cause. Opening that file does not require privileges but rather creation fails somewhere else but OpenVPN does not report this. I should be able to continue with the investigation some time later this week. 

BTW, this is the strace of the failure:

```
openat(AT_FDCWD, "/dev/net/tun", O_RDWR) = -1 ENOENT (No such file or directory)
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0
write(1, "2022-07-06 13:34:28 ERROR: Canno"..., 1012022-07-06 13:34:28 ERROR: Cannot open TUN/TAP dev /dev/net/tun: No such file or directory (errno=2)
) = 101
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=118, ...}) = 0
write(1, "2022-07-06 13:34:28 Exiting due "..., 472022-07-06 13:34:28 Exiting due to fatal error
) = 47
exit_group(1)                           = ?
+++ exited with 1 +++
```
That system call does fail if the file does not exist.

**Comment by ialidzhikov** [comment]
I can also share that the [vpn-seed-server/vpn-seed-server](https://github.com/gardener/gardener/blob/v1.50.1/pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go#L350-L357) fails in a similar way when I try to run it in unprivileged mode.

```
2022-07-15 12:19:16 WARNING: --topology net30 support for server configs with IPv4 pools will be removed in a future release. Please migrate to --topology subnet as soon as possible.
2022-07-15 12:19:16 DEPRECATED OPTION: --cipher set to 'AES-256-CBC' but missing in --data-ciphers (AES-256-GCM:AES-128-GCM). Future OpenVPN version will ignore --cipher for cipher negotiations. Add 'AES-256-CBC' to --data-ciphers or change --cipher 'AES-256-CBC' to --data-ciphers-fallback 'AES-256-CBC' to silence this warning.
2022-07-15 12:19:16 WARNING: file '/srv/secrets/vpn-server/tls.key' is group or others accessible
2022-07-15 12:19:16 OpenVPN 2.5.6 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Mar 24 2022
2022-07-15 12:19:16 library versions: OpenSSL 1.1.1n  15 Mar 2022, LZO 2.10
2022-07-15 12:19:16 NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
2022-07-15 12:19:16 ERROR: Cannot open TUN/TAP dev /dev/net/tun: No such file or directory (errno=2)
2022-07-15 12:19:16 Exiting due to fatal error
```

I can confirm that for an unprivileged container the `/dev/net/tun` device is missing while for a privileged one it is present (I assume it's the host device?):
```
$ k exec -ti vpn-seed-server-unprivileged -- ls -la /dev/net/tun
ls: /dev/net/tun: No such file or directory
command terminated with exit code 1

$ k exec -ti vpn-seed-server-privileged -- ls -la /dev/net/tun
crw-rw-rw-    1 root     root       10, 200 Jul 18 06:28 /dev/net/tun
```

For now I don't know what's the use of this `/dev/net/tun` device and whether it really needs to mount it from the host or can it be something created in the container (ref https://community.openvpn.net/openvpn/wiki/UnprivilegedUser#TUNTAPDevice).
@ScheererJ @DockToFuture this week I will busy with GQF shift, but next week we can sync and check how we can fix this issue.

**Comment by marwinski** [comment]
I am not sure that is the problem... I have an appointment with an OpenVPN maintainer tomorrow to look at this ;-)

**Comment by vlerenc** [comment]
@marwinski Then, maybe it helps/motivates the maintainer(s) to understand that privileged containers are not permitted (at all) by [CIS](https://www.cisecurity.org/benchmark/kubernetes) and [DISA STIG K8s](https://stigviewer.com/stig/kubernetes/2021-04-14/). They do not have to care obviously (and OpenVPN is not bound to K8s as Calico is), but maybe it shows that this is somewhat relevant nowadays for security reasons by commercial and official standards.

**Comment by marwinski** [comment]
actually, they do but just started recently ...

**Comment by ScheererJ** [comment]
https://github.com/gardener/gardener/pull/6346 addresses the topic for the reversed vpn server side, which is surprisingly easy (just removing the privileged flag).

**Comment by ScheererJ** [comment]
https://github.com/gardener/gardener/pull/6349 and https://github.com/gardener/gardener/pull/6352 address the topic for the reversed vpn client side, which is unfortunately a bit more complicated.

**Comment by marwinski** [comment]
Found the issue: when a docker container run privileged docker creates the character device `/dev/net/tun` which is a prerequisite for OpenVPN. Docker does not create the character device when run in an unprivileged setup. The fix is to create the device before starting OpenVPN (see PR).

**Comment by vlerenc** [comment]
Nice @marwinski, 👍.

**Comment by ialidzhikov** [comment]
@marwinski you closed https://github.com/gardener/vpn2/pull/13. If the way forward is to mount `/dev/net/tun` from the host, please review https://github.com/gardener/gardener/pull/6346 and https://github.com/gardener/gardener/pull/6352. Thanks in advance!

**Comment by ialidzhikov** [comment]
I think we can close this issue after https://github.com/gardener/gardener/pull/6346 and https://github.com/gardener/gardener/pull/6352 are merged. Thank you all for the help, appreciated!

@vlerenc we don't want to invest in the old vpn-shoot client (that gets deployed when the `ReversedVPN` feature gate is disabled), right? For now it still runs in privileged mode, but the `ReversedVPN` is beta (enabled by default) and I assume the mid-term plan of the team is to drop the old vpn solution.

**Comment by vlerenc** [comment]
@ialidzhikov Correct.
